File: scripts/regenerate-audio.js
const fs = require('fs');
const path = require('path');
const db = require('../src/database/db');
const tts = require('../src/rag/gemini-tts');
const REPORT = 'D:\\local-rag-voice-bot\\speed-math-master\\audio\\speech\\wav-quality-report.json';
const CACHE = 'D:\\local-rag-voice-bot\\speed-math-master\\audio\\speech\\cache';
const VISEMES = 'D:\\local-rag-voice-bot\\speed-math-master\\audio\\speech\\cache\\visemes';

async function main() {
  console.log('=== Audio Regeneration ===');
  if (!tts.available) { console.error('No API keys!'); process.exit(1); }
  console.log('API keys:', tts.apiKeys.length);

  const report = JSON.parse(fs.readFileSync(REPORT, 'utf8'));
  const padding = report.padding;
  const mismatch = report.mismatch;
  const allIssues = [...padding, ...mismatch];
  console.log('Issues:', allIssues.length, '(padded:', padding.length, 'mismatch:', mismatch.length, ')');

  const exRes = await db.query('SELECT source_id, hint_text, speech_text, quick_trick FROM exercises');
  const map = {};
  for (const ex of exRes.rows) map[ex.source_id] = ex;

  let ok = 0, fail = 0, skip = 0;
  for (let i = 0; i < allIssues.length; i++) {
    const item = allIssues[i];
    const src = item.src || item.sourceId;
    const kind = item.kind || (item.f.includes('_hint') ? 'hint' : 'trick');
    const baseName = src + '_' + kind;
    const ex = map[src];
    if (!ex) { skip++; continue; }
    const text = kind === 'hint' ? (ex.hint_text || ex.speech_text) : (ex.quick_trick || ex.speech_text);
    if (!text) { skip++; continue; }
    try {
      const { audioBuffer, sampleRate } = await tts.synthesize(text);
      const wav = tts.pcmToWav(audioBuffer, sampleRate);
      fs.writeFileSync(path.join(CACHE, baseName + '.wav'), wav);
      const vis = tts.pcmToVisemes(audioBuffer, sampleRate);
      if (vis) fs.writeFileSync(path.join(VISEMES, baseName + '.json'), JSON.stringify(vis));
      ok++;
      if (ok % 50 === 0) console.log('  Progress:', ok, 'ok', fail, 'fail', skip, 'skip');
    } catch (e) {
      fail++;
      console.log('  FAIL:', baseName, e.message.substring(0, 80));
    }
  }
  console.log('\n=== DONE ===');
  console.log('OK:', ok, 'FAIL:', fail, 'SKIP:', skip);
  await db.end().catch(() => {});
}
main().catch(e => { console.error(e); process.exit(1); });

File: scripts/regenerate_audio.py
import json, asyncio, edge_tts, os, sys, time
from pathlib import Path

CACHE = Path(r"D:\local-rag-voice-bot\speed-math-master\audio\speech\cache")
REPORT = Path(r"D:\local-rag-voice-bot\speed-math-master\audio\speech\wav-quality-report.json")
EXERCISES = Path(r"D:\local-rag-voice-bot\cadas-app-backend\tmp-exercises.json")
VOICE = "id-ID-GadisNeural"
MAX_RETRIES = 3
RETRY_DELAY = 2

report = json.loads(REPORT.read_text())
exercises = json.loads(EXERCISES.read_text())
mp = {e["source_id"]: e for e in exercises}
issues = report.get("padding", []) + report.get("mismatch", [])
print(f"Issues: {len(issues)}, DB rows: {len(mp)}")

async def gen_one(text, out_path):
    for attempt in range(MAX_RETRIES):
        try:
            comm = edge_tts.Communicate(text, VOICE)
            await comm.save(str(out_path))
            return True
        except Exception as e:
            if attempt < MAX_RETRIES - 1:
                await asyncio.sleep(RETRY_DELAY * (attempt + 1))
            else:
                return False

async def main():
    ok = fail = skip = 0
    for i, item in enumerate(issues):
        src = item.get("src") or item.get("sourceId")
        kind = item.get("kind") or ("hint" if "_hint" in item.get("f", "") else "trick")
        base = f"{src}_{kind}"
        ex = mp.get(src)
        if not ex:
            skip += 1
            continue
        text = ex.get("hint_text") or ex.get("speech_text") if kind == "hint" else ex.get("quick_trick") or ex.get("speech_text")
        if not text:
            skip += 1
            continue
        out = CACHE / f"{base}.wav"
        success = await gen_one(text, out)
        if success:
            ok += 1
        else:
            fail += 1
            if fail <= 5:
                print(f"  FAIL: {base}")
        if ok % 50 == 0:
            print(f"  Progress: {ok} ok, {fail} fail, {skip} skip")
        await asyncio.sleep(0.1)
    print(f"\n=== DONE ===")
    print(f"OK: {ok}, FAIL: {fail}, SKIP: {skip}")

asyncio.run(main())
